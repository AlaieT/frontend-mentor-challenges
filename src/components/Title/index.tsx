import React from "react";

import type { TitleProps } from "../../types";

import styles from "../../styles/components/title.module.scss";

const Title = ({ sentence, delay }: TitleProps) => {
  const splitText = React.useMemo(() => Array.from(sentence), [sentence]);
  const [count, setCount] = React.useState(-1);

  React.useEffect(() => {
    const interval = setInterval(() => {
      setCount((item) => {
        if (item === sentence.length - 1) {
          clearInterval(interval);
          return item;
        }
        return item + 1;
      });
    }, delay);
    return () => clearInterval(interval);
  }, [sentence, splitText, delay]);

  return (
    <h1 id={styles.title}>
      {splitText.map((letter, idx) => (
        <span
          key={`title_${letter + idx}`}
          className={idx > count ? styles.disabled : undefined}
        >
          {letter === " " ? <>&nbsp;</> : letter}
        </span>
      ))}
    </h1>
  );
};

export default Title;
