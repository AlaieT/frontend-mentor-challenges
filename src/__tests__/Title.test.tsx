import React from "react";
import { render, cleanup, act } from "@testing-library/react";

import Title from "../components/Title";

afterEach(cleanup);

describe("Title", () => {
  const props = {
    sentence: "frontend mentorâ€”challenges",
    delay: 10
  };

  describe("correct render", () => {
    it("should match snapshot before animation", () => {
      expect(
        render(<Title sentence={props.sentence} delay={props.delay} />).asFragment()
      ).toMatchSnapshot();
    });

    it("should match snapshot after animation", async () => {
      const { asFragment } = render(<Title sentence={props.sentence} delay={props.delay} />);

      await act(async () => {
        await new Promise((resolve, _) => {
          setTimeout(() => resolve("done"), props.sentence.length * props.delay + 50);
        });
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
