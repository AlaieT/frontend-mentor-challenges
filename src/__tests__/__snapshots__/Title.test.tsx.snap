// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title correct render should match snapshot after animation 1`] = `
<DocumentFragment>
  <h1
    id="title"
  >
    <span>
      f
    </span>
    <span>
      r
    </span>
    <span>
      o
    </span>
    <span>
      n
    </span>
    <span>
      t
    </span>
    <span>
      e
    </span>
    <span>
      n
    </span>
    <span>
      d
    </span>
    <span>
       
    </span>
    <span>
      m
    </span>
    <span>
      e
    </span>
    <span>
      n
    </span>
    <span>
      t
    </span>
    <span>
      o
    </span>
    <span>
      r
    </span>
    <span>
      —
    </span>
    <span>
      c
    </span>
    <span>
      h
    </span>
    <span>
      a
    </span>
    <span>
      l
    </span>
    <span>
      l
    </span>
    <span>
      e
    </span>
    <span>
      n
    </span>
    <span>
      g
    </span>
    <span>
      e
    </span>
    <span>
      s
    </span>
  </h1>
</DocumentFragment>
`;

exports[`Title correct render should match snapshot before animation 1`] = `
<DocumentFragment>
  <h1
    id="title"
  >
    <span
      class="disabled"
    >
      f
    </span>
    <span
      class="disabled"
    >
      r
    </span>
    <span
      class="disabled"
    >
      o
    </span>
    <span
      class="disabled"
    >
      n
    </span>
    <span
      class="disabled"
    >
      t
    </span>
    <span
      class="disabled"
    >
      e
    </span>
    <span
      class="disabled"
    >
      n
    </span>
    <span
      class="disabled"
    >
      d
    </span>
    <span
      class="disabled"
    >
       
    </span>
    <span
      class="disabled"
    >
      m
    </span>
    <span
      class="disabled"
    >
      e
    </span>
    <span
      class="disabled"
    >
      n
    </span>
    <span
      class="disabled"
    >
      t
    </span>
    <span
      class="disabled"
    >
      o
    </span>
    <span
      class="disabled"
    >
      r
    </span>
    <span
      class="disabled"
    >
      —
    </span>
    <span
      class="disabled"
    >
      c
    </span>
    <span
      class="disabled"
    >
      h
    </span>
    <span
      class="disabled"
    >
      a
    </span>
    <span
      class="disabled"
    >
      l
    </span>
    <span
      class="disabled"
    >
      l
    </span>
    <span
      class="disabled"
    >
      e
    </span>
    <span
      class="disabled"
    >
      n
    </span>
    <span
      class="disabled"
    >
      g
    </span>
    <span
      class="disabled"
    >
      e
    </span>
    <span
      class="disabled"
    >
      s
    </span>
  </h1>
</DocumentFragment>
`;
